IdentityCore/src/throttling/cache/MSIDThrottlingCacheService.m

```diff
@@ -0,0 +1,364 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.  
+
+
+#import <Foundation/Foundation.h>
+#import "MSIDThrottlingCacheService.h"
+#import "MSIDThrottlingCacheNode.h"
+#import "MSIDThrottlingCacheRecord.h"
+#import "MSIDCache.h"
+
+static NSString *const HEAD_NODE_KEY = @"HEAD";
+static NSString *const TAIL_NODE_KEY = @"TAIL";
+
+@interface MSIDThrottlingCacheNode ()
+
+- (void)setPrevRequestThumbprintKey:(NSString *)prevRequestThumbprintKey;
+- (void)setNextRequestThumbprintKey:(NSString *)nextRequestThumbprintKey;
+
+@end
+
+@interface MSIDThrottlingCacheService ()
+
+@property (nonatomic) NSUInteger cacheSizeInt;
+@property (nonatomic) MSIDThrottlingCacheNode *head;
+@property (nonatomic) MSIDThrottlingCacheNode *tail;
+@property (nonatomic) NSMutableDictionary *container;
+@property (nonatomic) dispatch_queue_t synchronizationQueue;
+
+@end
+
+@implementation MSIDThrottlingCacheService
+
+- (NSUInteger)cacheSize
+{
+    return self.cacheSizeInt-2;
+}
+
+- (NSUInteger)numCacheRecords
+{
+    return self.container.allKeys.count-2;
+}
+
+- (instancetype)initWithThrottlingCacheSize:(NSUInteger)cacheSize
+{
+    self = [super init];
+    if (self)
+    {
+        _cacheSizeInt = cacheSize+2;
+        //create dummy head and tail
+        _head = [[MSIDThrottlingCacheNode alloc] initWithThumbprintKey:HEAD_NODE_KEY
+                                                         errorResponse:nil
+                                                          throttleType:nil
+                                                      throttleDuration:0
+                                              prevRequestThumbprintKey:nil
+                                              nextRequestThumbprintKey:TAIL_NODE_KEY];
+        
+        _tail = [[MSIDThrottlingCacheNode alloc] initWithThumbprintKey:TAIL_NODE_KEY
+                                                         errorResponse:nil
+                                                          throttleType:nil
+                                                      throttleDuration:0
+                                              prevRequestThumbprintKey:HEAD_NODE_KEY
+                                              nextRequestThumbprintKey:nil];
+        
+        //create concurrent queue and container dictionary
+        NSString *queueName = [NSString stringWithFormat:@"com.microsoft.msidthrottlingcache-%@", [NSUUID UUID].UUIDString];
+        _synchronizationQueue = dispatch_queue_create([queueName cStringUsingEncoding:NSASCIIStringEncoding], DISPATCH_QUEUE_CONCURRENT);
+        _container = [NSMutableDictionary new];
+        
+        [self.container setObject:_head forKey:HEAD_NODE_KEY];
+        [self.container setObject:_tail forKey:TAIL_NODE_KEY];
+        
+        
+    }
+    return self;
+}
+
++ (MSIDThrottlingCacheService *)sharedInstance:(NSUInteger)cacheSize
```

Comment:
cacheSize is ignored here for subsequent calls to sharedInstance. I think you should use a default cacheSize instead of passing it as a parameter.

Reply:
good catch.

Reply:
How can we determine the cacheSize value here?

Reply:
as Olga suggested, I will create a default cacheSize for shared instance. if the calling app wants a different cache size, they can use the init method directly.